From fccc7f7b31f6f08713db8adff94f84761bd441bf Mon Sep 17 00:00:00 2001
From: Sergio Villar Senin <svillar@igalia.com>
Date: Tue, 7 Feb 2023 10:13:47 +0100
Subject: [PATCH 10/16] Fix for DeligthVR video player

Immersive videos played with the DelightVR player fail to play the
very first time if no previous WebXR experience has been loaded.
More technically speaking they fail if the XR devices were not
previously enumerated.

The device enumeration is normally triggered by the XRSystem once
it gets the permission to start an XRSession. However we have
noticed that this does not work in some cases, like the DelightVR
player used to play immersive videos in tons of sites. WebXR
applications issue permission requests as the webpage is loaded,
meaning that by the time the user clicks on EnterVR-like button
the devices are already enumerated. However in the case of
DelightVR that only happens when the user already clicks on play.
For some unknown reason, that's too late, probably the WebGL
context is not properly already assigned to the newly enumerated
device (maybe this https://bugzilla.mozilla.org/show_bug.cgi?id=1580258).

So the hacky solution is to perform that enumeration on advance.
That should not be a problem in practice because we only do that
after successfully detecting that a runtime is present, meaning
that we do that only for web pages already using WebXR API. It is
just it isn't the most correct way to do that (it's better to do
it only after getting the OK from the permission request). In any
case this does not break the API contract as far as I understand
because the specs only mention that tracking should not be started
until really needed, but here we're just listing the devices.
---
 gfx/vr/VRManager.cpp | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/gfx/vr/VRManager.cpp b/gfx/vr/VRManager.cpp
index 2157439b92b1..e70a04fc2b55 100644
--- a/gfx/vr/VRManager.cpp
+++ b/gfx/vr/VRManager.cpp
@@ -567,6 +567,15 @@ void VRManager::DetectRuntimes() {
     // after installing or removing an XR device
     // runtime.
     DispatchRuntimeCapabilitiesUpdate();
+    // FIXME: this should not be done here. The device enumeration is normally triggered
+    // by the XRSystem once it gets the permission to start an XRSession. However we have
+    // noticed that this does not work in some cases, like the DelightVR player used to
+    // play immersive videos in tons of sites. WebXR applications issue permission requests
+    // as the webpage is loaded, meaning that by the time the user clicks on EnterVR-like
+    // button the devices are already enumerated. However in the case of DelightVR that only
+    // happens when the user already clicks on play. For some unknown reason, that's too late,
+    // probably the WebGL context is not properly already assigned to the newly enumerated device.
+    EnumerateDevices();
     return;
   }
   mRuntimeDetectionRequested = true;
-- 
2.35.1

